src/Adapter/MSTestAdapter.PlatformServices/Services/ThreadOperations.cs

```diff
@@ -20,7 +24,38 @@ public class ThreadOperations : IThreadOperations
     public bool Execute(Action action, int timeout, CancellationToken cancelToken)
     {
 #if NETFRAMEWORK
-        if (cancelToken.IsCancellationRequested)
+        return ExecuteWithThread(action, timeout, cancelToken);
+#else
+        return RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && Thread.CurrentThread.GetApartmentState() == ApartmentState.STA
+            ? ExecuteWithThread(action, timeout, cancelToken)
+            : ExecuteWithThreadPool(action, timeout, cancelToken);
+#endif
+    }
+
+#if !NETFRAMEWORK
+    private static bool ExecuteWithThreadPool(Action action, int timeout, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var executionTask = Task.Run(action, cancellationToken);
+
+            // False means execution timed out.
+            return executionTask.Wait(timeout, cancellationToken);
+        }
+        catch (OperationCanceledException)
```

Comment:
I would verify the ct in the catch == cancellationToken.Token

Reply:
ok, wonder if in future a log that report the "wrong" token caould be useful in case of some but let's move on as-is